### PR TITLE
[gh-457] add Schnorr signature support.

### LINKED ISSUE #457: [crypto] support `schnorr` signature standard for nostr and bitcoin protocol
Like `ECDSA` and `ED25519` signatures, we should also add supports for `Schnorr` signatures, verify the authenticity of `Schnorr` signatures, validate the signatures from the Rooch transaction validator in order to support crypto scheme validation for `Schnorr` in Move.

At the time of writing, nostr mainly uses `Schnorr` signatures for its events sig, public key, and encodings as described here:

https://github.com/nostr-protocol/nips/blob/7cd861c4d3c7a28d6b0d4ee26ad2a36fdb2670a5/01.md#events-and-signatures

Bitcoin also has proposal `BIP-340` to support `Schnorr` signatures as mentioned here:

https://bips.xyz/340

As a practical example is being built for nostr, there shall be added a crypto verification for `Schnorr` signatures.

Part of #406

### PR DESCRIPTION
**Pull request by feliciss**
This PR resolves #457

### CODE REVIEW

**Comment by feliciss** [comment]
I would like to remove traits imported from `fastcrypto` and generalize trait `RoochSignatureInner` and use it for SchnorrRoochSignature, Ed25519RoochSignature and ECDSARoochSignature.

But I encountered some difficult issues:

1. To satisfy `AsRef<[u8]>` of RoochAddress:

```zsh
error[E0277]: the trait bound `secp256k1::XOnlyPublicKey: AsRef<[u8]>` is not satisfied
   --> crates/rooch-types/src/crypto.rs:812:48
    |
812 |         let received_addr = RoochAddress::from(&pk);
    |                             ------------------ ^^^ the trait `AsRef<[u8]>` is not implemented for `secp256k1::XOnlyPublicKey`
    |                             |
    |                             required by a bound introduced by this call
    |
    = help: the following other types implement trait `From<T>`:
              <address::RoochAddress as From<&T>>
              <address::RoochAddress as From<&crypto::PublicKey>>
              <address::RoochAddress as From<move_core_types::account_address::AccountAddress>>
note: required for `address::RoochAddress` to implement `From<&secp256k1::XOnlyPublicKey>`
```

2. Duplicate implementation of `Secp256k1KeyPair `

```rs
// TODO generalize Secp256k1KeyPair
// impl Signer<Signature> for Secp256k1KeyPair {
//     fn sign(&self, msg: &[u8]) -> Signature {
//         ECDSARoochSignature::new(self, msg).into()
//     }
// }
```

```rs
impl Signer<Signature> for Secp256k1KeyPair {
    fn sign(&self, msg: &[u8]) -> Signature {
        SchnorrRoochSignature::new(self, msg).into()
    }
}
```

They are using same key pair base for `ECDSA` and `Schnorr` signatures.

Other wreaked issues:

```zsh
error[E0277]: the trait bound `OsRng: rand_core::CryptoRng` is not satisfied
   --> crates/rooch-types/src/transaction/authenticator.rs:60:50
    |
60  |         let keypair: Keypair = Keypair::generate(&mut csprng);
    |                                ----------------- ^^^^^^^^^^^ the trait `rand_core::CryptoRng` is not implemented for `OsRng`
    |                                |
    |                                required by a bound introduced by this call
    |
    = help: the following other types implement trait `rand_core::CryptoRng`:
              &'a mut R
              Box<R>
              rand::rngs::adapter::reseeding::ReseedingRng<R, Rsdr>
              rand::rngs::entropy::EntropyRng
              rand::rngs::std::StdRng
              rand::rngs::thread::ThreadRng
              rand_chacha::chacha::ChaCha12Core
              rand_chacha::chacha::ChaCha12Rng
            and 6 others
note: required by a bound in `ed25519_dalek::Keypair::generate`
    |
129 |         R: CryptoRng + RngCore,
    |            ^^^^^^^^^ required by this bound in `Keypair::generate`
```

```zsh
error[E0277]: the trait bound `StdRng: crypto::AllowedRng` is not satisfied
    --> crates/rooch-types/src/crypto.rs:1045:27
     |
1045 |     let kp = KP::generate(&mut StdRng::from_rng(csprng).unwrap());
     |              ------------ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `crypto::AllowedRng` is not implemented for `StdRng`
     |              |
     |              required by a bound introduced by this call
     |
note: required by a bound in `KeypairTraits::generate`
    --> crates/rooch-types/src/crypto.rs:440:20
     |
440  |     fn generate<R: AllowedRng>(rng: &mut R) -> Self;
     |                    ^^^^^^^^^^ required by this bound in `KeypairTraits::generate`
```

I'm still solving the above problems, but I might need some reviews and hints. @jolestar @wow-sven 

There're conflicts between #466 and #461, mainly for renaming `spec256k1` to `ECDSA` in `crates/rooch-types/src/transaction/authenticator.rs`.

**Comment by wow-sven** [comment]
> 


> I would like to remove traits imported from `fastcrypto` and generalize trait `RoochSignatureInner` and use it for SchnorrRoochSignature, Ed25519RoochSignature and ECDSARoochSignature.
> 
> But I encountered some difficult issues:
> 
> 1. To satisfy `AsRef<[u8]>` of RoochAddress:
> 
> ```shell
> error[E0277]: the trait bound `secp256k1::XOnlyPublicKey: AsRef<[u8]>` is not satisfied
>    --> crates/rooch-types/src/crypto.rs:812:48
>     |
> 812 |         let received_addr = RoochAddress::from(&pk);
>     |                             ------------------ ^^^ the trait `AsRef<[u8]>` is not implemented for `secp256k1::XOnlyPublicKey`
>     |                             |
>     |                             required by a bound introduced by this call
>     |
>     = help: the following other types implement trait `From<T>`:
>               <address::RoochAddress as From<&T>>
>               <address::RoochAddress as From<&crypto::PublicKey>>
>               <address::RoochAddress as From<move_core_types::account_address::AccountAddress>>
> note: required for `address::RoochAddress` to implement `From<&secp256k1::XOnlyPublicKey>`
> ```
> 
>  2. Duplicate implementation of `Secp256k1KeyPair `
> 
> ```rust
> // TODO generalize Secp256k1KeyPair
> // impl Signer<Signature> for Secp256k1KeyPair {
> //     fn sign(&self, msg: &[u8]) -> Signature {
> //         ECDSARoochSignature::new(self, msg).into()
> //     }
> // }
> ```
> 
>  ```rust
> impl Signer<Signature> for Secp256k1KeyPair {
>     fn sign(&self, msg: &[u8]) -> Signature {
>         SchnorrRoochSignature::new(self, msg).into()
>     }
> }
> ```
> 
> They are using same key pair base for `ECDSA` and `Schnorr` signatures.
> 
> Other wreaked issues:
> 
> ```shell
> error[E0277]: the trait bound `OsRng: rand_core::CryptoRng` is not satisfied
>    --> crates/rooch-types/src/transaction/authenticator.rs:60:50
>     |
> 60  |         let keypair: Keypair = Keypair::generate(&mut csprng);
>     |                                ----------------- ^^^^^^^^^^^ the trait `rand_core::CryptoRng` is not implemented for `OsRng`
>     |                                |
>     |                                required by a bound introduced by this call
>     |
>     = help: the following other types implement trait `rand_core::CryptoRng`:
>               &'a mut R
>               Box<R>
>               rand::rngs::adapter::reseeding::ReseedingRng<R, Rsdr>
>               rand::rngs::entropy::EntropyRng
>               rand::rngs::std::StdRng
>               rand::rngs::thread::ThreadRng
>               rand_chacha::chacha::ChaCha12Core
>               rand_chacha::chacha::ChaCha12Rng
>             and 6 others
> note: required by a bound in `ed25519_dalek::Keypair::generate`
>     |
> 129 |         R: CryptoRng + RngCore,
>     |            ^^^^^^^^^ required by this bound in `Keypair::generate`
> ```
> 
>  ```shell
> error[E0277]: the trait bound `StdRng: crypto::AllowedRng` is not satisfied
>     --> crates/rooch-types/src/crypto.rs:1045:27
>      |
> 1045 |     let kp = KP::generate(&mut StdRng::from_rng(csprng).unwrap());
>      |              ------------ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `crypto::AllowedRng` is not implemented for `StdRng`
>      |              |
>      |              required by a bound introduced by this call
>      |
> note: required by a bound in `KeypairTraits::generate`
>     --> crates/rooch-types/src/crypto.rs:440:20
>      |
> 440  |     fn generate<R: AllowedRng>(rng: &mut R) -> Self;
>      |                    ^^^^^^^^^^ required by this bound in `KeypairTraits::generate`
> ```
> 
> I'm still solving the above problems, but I might need some reviews and hints. @jolestar @wow-sven
> 
> There're conflicts between #466 and #461, mainly for renaming `spec256k1` to `ECDSA` in `crates/rooch-types/src/transaction/authenticator.rs`.

Option1: 
Keep making changes based on this commit （865e9dce759a71a729f5e0f527e19e7dbc209a5e） without breaking the current Fastcrypto, can be pushed to Fastcrypto

Fork fastcrypto
- Implements the Authenticator trait for SchnorrSignature
- Implements the VerifyingKey trait for XOnlyPublicKey
- Implements the fastcrypto::KeyPair trait for KeyPair

ref：https://github.com/MystenLabs/fastcrypto/issues/51

Option2:
Remove fastcrypto completely and reorganize instead of keeping parts. Avoid further incompatibilities later on

**Comment by feliciss** [comment]
> > I would like to remove traits imported from `fastcrypto` and generalize trait `RoochSignatureInner` and use it for SchnorrRoochSignature, Ed25519RoochSignature and ECDSARoochSignature.
> > But I encountered some difficult issues:
> > 
> > 1. To satisfy `AsRef<[u8]>` of RoochAddress:
> > 
> > ```shell
> > error[E0277]: the trait bound `secp256k1::XOnlyPublicKey: AsRef<[u8]>` is not satisfied
> >    --> crates/rooch-types/src/crypto.rs:812:48
> >     |
> > 812 |         let received_addr = RoochAddress::from(&pk);
> >     |                             ------------------ ^^^ the trait `AsRef<[u8]>` is not implemented for `secp256k1::XOnlyPublicKey`
> >     |                             |
> >     |                             required by a bound introduced by this call
> >     |
> >     = help: the following other types implement trait `From<T>`:
> >               <address::RoochAddress as From<&T>>
> >               <address::RoochAddress as From<&crypto::PublicKey>>
> >               <address::RoochAddress as From<move_core_types::account_address::AccountAddress>>
> > note: required for `address::RoochAddress` to implement `From<&secp256k1::XOnlyPublicKey>`
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > 2. Duplicate implementation of `Secp256k1KeyPair `
> > 
> > ```rust
> > // TODO generalize Secp256k1KeyPair
> > // impl Signer<Signature> for Secp256k1KeyPair {
> > //     fn sign(&self, msg: &[u8]) -> Signature {
> > //         ECDSARoochSignature::new(self, msg).into()
> > //     }
> > // }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```rust
> > impl Signer<Signature> for Secp256k1KeyPair {
> >    fn sign(&self, msg: &[u8]) -> Signature {
> >        SchnorrRoochSignature::new(self, msg).into()
> >    }
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > They are using same key pair base for `ECDSA` and `Schnorr` signatures.
> > Other wreaked issues:
> > ```shell
> > error[E0277]: the trait bound `OsRng: rand_core::CryptoRng` is not satisfied
> >    --> crates/rooch-types/src/transaction/authenticator.rs:60:50
> >     |
> > 60  |         let keypair: Keypair = Keypair::generate(&mut csprng);
> >     |                                ----------------- ^^^^^^^^^^^ the trait `rand_core::CryptoRng` is not implemented for `OsRng`
> >     |                                |
> >     |                                required by a bound introduced by this call
> >     |
> >     = help: the following other types implement trait `rand_core::CryptoRng`:
> >               &'a mut R
> >               Box<R>
> >               rand::rngs::adapter::reseeding::ReseedingRng<R, Rsdr>
> >               rand::rngs::entropy::EntropyRng
> >               rand::rngs::std::StdRng
> >               rand::rngs::thread::ThreadRng
> >               rand_chacha::chacha::ChaCha12Core
> >               rand_chacha::chacha::ChaCha12Rng
> >             and 6 others
> > note: required by a bound in `ed25519_dalek::Keypair::generate`
> >     |
> > 129 |         R: CryptoRng + RngCore,
> >     |            ^^^^^^^^^ required by this bound in `Keypair::generate`
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```shell
> > error[E0277]: the trait bound `StdRng: crypto::AllowedRng` is not satisfied
> >    --> crates/rooch-types/src/crypto.rs:1045:27
> >     |
> > 1045 |     let kp = KP::generate(&mut StdRng::from_rng(csprng).unwrap());
> >     |              ------------ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `crypto::AllowedRng` is not implemented for `StdRng`
> >     |              |
> >     |              required by a bound introduced by this call
> >     |
> > note: required by a bound in `KeypairTraits::generate`
> >    --> crates/rooch-types/src/crypto.rs:440:20
> >     |
> > 440  |     fn generate<R: AllowedRng>(rng: &mut R) -> Self;
> >     |                    ^^^^^^^^^^ required by this bound in `KeypairTraits::generate`
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I'm still solving the above problems, but I might need some reviews and hints. @jolestar @wow-sven
> > There're conflicts between #466 and #461, mainly for renaming `spec256k1` to `ECDSA` in `crates/rooch-types/src/transaction/authenticator.rs`.
> 
> Duplicate implementation of `Secp256k1KeyPair `
> 
> Option1: Keep making changes based on this commit （865e9dce759a71a729f5e0f527e19e7dbc209a5e） without breaking the current Fastcrypto, can be pushed to Fastcrypto
> 
> Fork fastcrypto, Add schnorr to it.
> 
> * Implements the Authenticator trait for SchnorrSignature
> * Implements the VerifyingKey trait for XOnlyPublicKey
> * Implements the fastcrypto::KeyPair trait for KeyPair
> 
> ref：[MystenLabs/fastcrypto#51](https://github.com/MystenLabs/fastcrypto/issues/51)
> 
> Option2: Remove fastcrypto completely and reorganize instead of keeping parts. Avoid further incompatibilities later on

Thanks. 🙂 I'll do Option 2 to decouple it from fastcrypto. I don't think it's a good idea to keep dependencies on commercial libraries.

**Comment by jolestar** [comment]
[fastcrypto](https://github.com/MystenLabs/fastcrypto) is in Apache2/MIT license. I think Option 1 is better.

**Comment by feliciss** [comment]
> [fastcrypto](https://github.com/MystenLabs/fastcrypto) is in Apache2/MIT license. I think Option 1 is better.

I can add Option 1 temporarily, but in the long term view, it may needs to decouple from fastcrypto if that repo isn't being maintained.

**Comment by feliciss** [comment]
> [fastcrypto](https://github.com/MystenLabs/fastcrypto) is in Apache2/MIT license. I think Option 1 is better.

Can you @wow-sven @jolestar make a fork the repo [fastcrypto](https://github.com/MystenLabs/fastcrypto)? I do not have access to the team account rooch-network on GitHub.

**Comment by jolestar** [comment]
> > [fastcrypto](https://github.com/MystenLabs/fastcrypto) is in Apache2/MIT license. I think Option 1 is better.
> 
> Can you @wow-sven @jolestar make a fork the repo [fastcrypto](https://github.com/MystenLabs/fastcrypto)? I do not have access to the team account rooch-network on GitHub. Or I can fork to my own and submit it later to the team repo.

https://github.com/rooch-network/fastcrypto

**Comment by feliciss** [comment]
@wow-sven @jolestar I can almost finish it, but, there's another duplicate implementation. Do you have any idea to resolve it?

Code:

```rs
#[derive(Debug, Clone, PartialEq, Eq, JsonSchema)]
// TODO From here
// #[derive(Debug, Clone, PartialEq, Eq, From, JsonSchema)]
pub enum PublicKey {
    Ed25519(Ed25519PublicKeyAsBytes),
    Ecdsa(Secp256k1PublicKeyAsBytes),
    Schnorr(SchnorrPublicKeyAsBytes),
}

// TODO resolve From conflicts
impl From<BytesRepresentation<32>> for PublicKey {
    fn from(pk: BytesRepresentation<32>) -> Self {
        PublicKey::Ed25519(pk)
    }
}

impl AsRef<[u8]> for PublicKey {
    fn as_ref(&self) -> &[u8] {
        match self {
            PublicKey::Ed25519(pk: &BytesRepresentation<32>) => &pk.0,
            PublicKey::Ecdsa(pk: &BytesRepresentation<33>) => &pk.0,
            PublicKey::Schnorr(pk: &BytesRepresentation<32>) => &pk.0,
        }
    }
}
```

There're two hidden implementation of `&BytesRepresentation<32>` here.

**Comment by feliciss** [comment]
It's ready for review now.

File: crates/rooch-framework/src/natives/rooch_framework/crypto/schnorr.rs
Comment (jolestar) [inline crates/rooch-framework/src/natives/rooch_framework/crypto/schnorr.rs:1]: Need a Move file to export the native function

File: Cargo.toml [hunk 1]
```diff
@@ -109,7 +109,7 @@ enum_dispatch = "^0.3"
 ethereum-types = "0.14.1"
 ethers = { version = "2.0.7", features = ["legacy"] }
 eyre = "0.6.8"
-fastcrypto = { git = "https://github.com/MystenLabs/fastcrypto", rev = "c27e87fb539c88b6fe1eac7dd82561254bb1e07a" }
+fastcrypto = { git = "https://github.com/feliciss/fastcrypto", branch = "schnorr" }
```
Comment (jolestar) [inline Cargo.toml:112]: https://github.com/rooch-network/fastcrypto/pull/1 has merged, use `/rooch-network/fastcrypto`

File: crates/rooch-framework/src/natives/rooch_framework/crypto/schnorr.rs
Comment (feliciss) [reply crates/rooch-framework/src/natives/rooch_framework/crypto/schnorr.rs:1]: Added.

File: Cargo.toml (hunk 1 above)
Comment (feliciss) [reply Cargo.toml]: Resolved.

**Review by jolestar — APPROVED** [review]

**Pull request merged**